### ISSUE TITLE
Summary and Climatic summary dialogue

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/R-Instat#4106]
The **Prepare > Column; Reshape > Summaries** has been used a lot and a variety of issues have arisen.

- [x] 1. A simple one to fix is that the **Omit Missing Values** checkbox is omitted on the **Climatic Summaries** dialogue.  That made it unusable for us.  It should be easy to add to the options.
- [ ] 2. There is no option to choose the **name of the summary data frame.**  This was requested.  That needs a bit of thought, because you could then choose a different name on subsequent uses of the dialogue - then what of the linking.  But I can see the value, not least because the current names are very long - more like labels!
- [ ] 3. There is also no option to choose the names of the resulting columns.  That's difficult to change, because of the range of columns that can be produced at one go.  But what is a concern is that getting the same summary overwrites the previous copy.  (An example was the number of non-missing rain days, a) overall and then b) when filter to only have values > 0.) It could be good to solve that one below.  Then we would rarely need this.
- [ ] 4. I wonder if there could be a way of getting the number of rain days (for a user definition of rain) in the climatic summary, without doing a filter?  This could be solved by the next point.
- [ ] 5. I have no idea how to use the Proportions/Percentiles?  How would I get the 20%, and 80% point for example, or the proportion (and/or count) > certain levels.  The old Instat Stats > Describe dialogue shows what would be useful.

![image](https://user-images.githubusercontent.com/11226469/32078136-b607e778-ba9d-11e7-9940-a560c35ed362.png)

**Comment by dannyparsons** [comment]
@shadrackkibet could you do 1. urgently? That's simple and needed urgently for our next climatic workshop

**Comment by rdstern** [comment]
I now understand the proportions/percentages idea better.  But it isn't the same as the above request, which Danny says is just a simple addition to the Summaries sub-dialogue - probably another tab, say **Percentages**.

**Comment by rdstern** [comment]
I now have a more definite suggestion.  Danny tells me this is easy if we only do the calculation for a single value.  So let's do this first.
1) I suggest this can go **into a box** at the top of the **More** tab.  This moves all the existing summaries down and that is fine.
2) In the box there are 3 check-boxes, default unchecked.  The labels at **Percentile, Count, Proportion**.
3. If **percentile** is checked it opens a **text-box** into which the user types a number.  This is to be between 0 and 100. It could be 0, or 0.1, or 1, or 5, or 50 (median) or 99.99, or even 100.
4. If **Count** is checked then it opens 2 boxes.  The first gives the choice of **>, >=**, etc.  Same as in the Filter dialogue already.  Then there is a **text-box** into which a number is typed.
5. If **Proportion** is checked the same two fields are opened as for the Count.  There is a further **check-box** (default unchecked) with label **As Percentage**.
6. If percentile is checked, then this calls the **quantile** function - see the calculator.
7. If Count is checked, then the function is sum(logical), e.g. sum(x>35).
8. If Proportion is checked then it is the count divided by the number not missing.  If the percentage check-box is checked then the result is multiplied by 100.
9. I hope the dialogue is easy to construct.  I suggest the quantile (i.e. percentile) should be straightforward to implement.  The other 2 might need some ideas from David.

**Comment by shadrackkibet** [comment]
At the moment our summaries function can only allow us to do summaries for one column i.e `x`. To be able to implement **covariance** and **correlations** we required two columns ` x `and `y`. I think we should be able to do summaries of more than one column .This needs to be able to implemented.

_Alternatively,_
I am thinking of having two single receivers which in the main dialog or in the sdg (with a selector) only enabled if and only if these two checkboxes(**covariance** or **correlations**) are checked then we can run this functions independently from the main summaries function?  but in this we will not be able to do correlations with our factors as Roger wanted(i think!).This will produce single output at a time which we can have in our output window.

What do others thinks?

**Comment by rdstern** [comment]
My suggestion on cor and cov is that it is on a separate tab, possibly called bivariate.  This is so we can leave the rest of the dialogue unaffected.  This contains our usual selector, but restricted (as is done in filter) to the same data frame as is used on the main dialogue.
There is a single receiver.  This is the second variable in (say) the correlation.  Then the correlation is given as is any other summary, using all the factors and all the columns in turn in just the same way as any other summary

**Comment by dannyparsons** [comment]
I like this idea so that the main dialog is unchanged. It might be a bit awkward to use because you choose the columns to correlate in different places but I think this is the best option still.

**Comment by rdstern** [comment]
I have a further request.  But this needs discussion.  It is (for some summaries, particularly max and min to also allow carry columns.  Let me give an example.  I have daily temperature data for different years (e.g. Dodoma) and possibly also for different stations (e.g. our Ghana data).  I would like the hottest day each year, i.e. Tmax is the variable and the summary is the Max of Tmax in the year.  So that we can do already.  Now I also would like to know the date (or the day of year (doy) at which we had this Tmax.  

So, the max is on one particular row of the data (each year) and I would also like to be able to carry some other columns, e.g. the date column.  

I see this going on the same tab as the correlations and covariates.  It would be a multiple receiver (may want to carry more than one column.  There are other possible summaries that give this same possibility.  The most obvious are First, Last and nth.  I would be happy for a given use of the special summary dialogue (as opposed to the general system) to be just for a single summary.  So you choose one of the summaries (ordinary radio buttons) and then give the columns that are selected.
There is another possible issue - what happens (with max and min) if there are multiple dates with the same max (min) value.  My initial suggestion is that (in this case you could add a second column - perhaps logical, that is FALSE for any observation that is not unique, and is TRUE otherwise.  (And we just give the first occurrence of that value.)

**Comment by dannyparsons** [comment]
I think what you're asking for is a filter and then a subset of columns for cases like min and max, the same could be done for others as well I think. Is this the same dialog or something different?

**Comment by rdstern** [comment]
If you can filter to say (x = max(x)), or first(x) etc and then the dialogue specifies the subset of the columns to give, then yes.  Of course there may already be a more general filter in operation as well.

**Comment by dannyparsons** [comment]
I forgot there's also a `by` in there because it's `x == max(x)` within the factor levels you've selected, so it's very similar to the summaries. I'm not sure if it's less confusing to have them together or in separate dialogs.

**Comment by rdstern** [comment]
I have kept this and changed the milestone to 0.5.  But I think most of the points have probably now been addressed.  One point that remains can perhaps be dropped, namely:
4. I wonder if there could be a way of getting the number of rain days (for a user definition of rain) in the climatic summary, without doing a filter? This could be solved by the next point.

This could still perhaps be useful.  But I found it quite useful in teaching - and not too difficult to do - to first do a simple calculation to get a logical column of rain>0.85 and then the climatic calculator just gets the sum of this column.
